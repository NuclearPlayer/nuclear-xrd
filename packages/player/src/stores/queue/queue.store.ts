import { create } from 'zustand';

import type { QueueItem, Track } from '@nuclearplayer/model';

import type { Queue, RepeatMode } from './queue.types';

type QueueStore = Queue & {
  loaded: boolean;
  loadFromDisk: () => Promise<void>;
  addToQueue: (tracks: Track[]) => void;
  addNext: (tracks: Track[]) => void;
  addAt: (tracks: Track[], index: number) => void;
  removeByIds: (ids: string[]) => void;
  removeByIndices: (indices: number[]) => void;
  clearQueue: () => void;
  reorder: (fromIndex: number, toIndex: number) => void;
  updateItemState: (id: string, updates: Partial<QueueItem>) => void;
  goToNext: () => void;
  goToPrevious: () => void;
  goToIndex: (index: number) => void;
  goToId: (id: string) => void;
  setRepeatMode: (mode: RepeatMode) => void;
  setShuffleEnabled: (enabled: boolean) => void;
  getCurrentItem: () => QueueItem | undefined;
};

export const useQueueStore = create<QueueStore>((set, get) => ({
  items: [],
  currentIndex: 0,
  repeatMode: 'off',
  shuffleEnabled: false,
  loaded: false,

  loadFromDisk: async () => {
    set({ loaded: true });
  },

  addToQueue: (tracks: Track[]) => {
    void tracks;
  },

  addNext: (tracks: Track[]) => {
    void tracks;
  },

  addAt: (tracks: Track[], index: number) => {
    void tracks;
    void index;
  },

  removeByIds: (ids: string[]) => {
    void ids;
  },

  removeByIndices: (indices: number[]) => {
    void indices;
  },

  clearQueue: () => {
    set({ items: [], currentIndex: 0 });
  },

  reorder: (fromIndex: number, toIndex: number) => {
    void fromIndex;
    void toIndex;
  },

  updateItemState: (id: string, updates: Partial<QueueItem>) => {
    void id;
    void updates;
  },

  goToNext: () => {
    const { currentIndex, items } = get();
    if (currentIndex < items.length - 1) {
      set({ currentIndex: currentIndex + 1 });
    }
  },

  goToPrevious: () => {
    const { currentIndex } = get();
    if (currentIndex > 0) {
      set({ currentIndex: currentIndex - 1 });
    }
  },

  goToIndex: (index: number) => {
    const { items } = get();
    if (index >= 0 && index < items.length) {
      set({ currentIndex: index });
    }
  },

  goToId: (id: string) => {
    const { items } = get();
    const index = items.findIndex((item) => item.id === id);
    if (index !== -1) {
      set({ currentIndex: index });
    }
  },

  setRepeatMode: (mode: RepeatMode) => {
    set({ repeatMode: mode });
  },

  setShuffleEnabled: (enabled: boolean) => {
    set({ shuffleEnabled: enabled });
  },

  getCurrentItem: () => {
    const { items, currentIndex } = get();
    return items[currentIndex];
  },
}));
