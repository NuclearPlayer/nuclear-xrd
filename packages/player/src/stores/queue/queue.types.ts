import type { QueueItem } from '@nuclearplayer/model';

export type RepeatMode = 'off' | 'all' | 'one';

export type Queue = {
  items: QueueItem[];
  currentIndex: number;
  repeatMode: RepeatMode;
  shuffleEnabled: boolean;
};
