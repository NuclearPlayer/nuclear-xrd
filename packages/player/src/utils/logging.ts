import * as Logger from '@tauri-apps/plugin-log';
import isError from 'lodash-es/isError';
import isString from 'lodash-es/isString';
import { toast } from 'sonner';

export const logFsError = async ({
  scope,
  command,
  targetPath,
  error,
  withToast,
  toastMessage,
}: {
  scope: string;
  command: string;
  targetPath: string;
  error: unknown; // Most often Error or string
  withToast?: boolean;
  toastMessage?: string;
}): Promise<void> => {
  const message = error instanceof Error ? error.message : String(error);

  if (withToast) {
    toast.error(toastMessage, { description: message });
  }

  await Logger.error(
    `[${scope}/fs] ${command} failed for ${targetPath}: ${message}`,
  );
};

export const resolveErrorMessage = (error: unknown): string => {
  if (isError(error)) {
    return error.message;
  }
  if (isString(error)) {
    return error;
  }
  return 'Unknown error';
};
