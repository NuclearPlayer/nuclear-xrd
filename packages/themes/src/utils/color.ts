import { converter, formatCss, oklch } from 'culori';

const asOklch = converter('oklch');

export type OklchComponents = { l: number; c: number; h: number };

export const parseOklch = (input: string): OklchComponents | null => {
  const { l, c, h } = oklch(input)!;
  return { l: l as number, c: c as number, h: h as number };
};

export const buildOklch = (components: OklchComponents): string =>
  formatCss({ mode: 'oklch', ...components });

export const replaceHue = (value: string, hue: number): string => {
  const { l, c } = asOklch(value)!;
  return formatCss({ mode: 'oklch', l, c, h: hue });
};
