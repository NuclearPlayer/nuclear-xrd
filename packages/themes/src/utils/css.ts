export const cssVarBlock = (
  selector: string,
  vars: Record<string, string>,
): string => {
  const lines = Object.entries(vars).map(
    ([key, value]) => `  --${key}: ${value};`,
  );
  return `${selector}{
${lines.join('\n')}
}`;
};
