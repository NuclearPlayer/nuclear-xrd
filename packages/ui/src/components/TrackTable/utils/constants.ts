export const DEFAULT_ROW_HEIGHT = 42;
export const DEFAULT_OVERSCAN = 8;
