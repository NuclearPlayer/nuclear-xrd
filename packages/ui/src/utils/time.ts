import { Duration } from 'luxon';

export const formatTime = (totalSeconds: number) => {
  const sign = totalSeconds < 0 ? '-' : '';
  const secs = Math.abs(Math.trunc(totalSeconds));
  const dur = Duration.fromObject({ seconds: secs }).shiftTo(
    'hours',
    'minutes',
    'seconds',
  );
  const hours = dur.hours ?? 0;
  const fmt = hours > 0 ? 'h:mm:ss' : 'm:ss';
  return `${sign}${dur.toFormat(fmt)}`;
};
