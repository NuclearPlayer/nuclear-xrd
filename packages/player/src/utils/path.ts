import { BaseDirectory, exists, mkdir } from '@tauri-apps/plugin-fs';

import { logFsError } from './logging';

export const ensureDir = async (
  dir: string,
  baseDir: BaseDirectory = BaseDirectory.AppData,
) => {
  try {
    const present = await exists(dir, { baseDir });
    if (!present) {
      try {
        await mkdir(dir, {
          recursive: true,
          baseDir,
        });
      } catch (e) {
        await logFsError('ensureDir', 'fs.mkdir', dir, e);
      }
    }
  } catch (e) {
    await logFsError('ensureDir', 'fs.exists', dir, e);
  }
  return dir;
};
