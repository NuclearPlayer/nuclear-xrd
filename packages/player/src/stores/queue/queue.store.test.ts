import '../../test/setup';

import { beforeEach, describe, expect, it, vi } from 'vitest';

import type { Track } from '@nuclearplayer/model';

import { resetInMemoryTauriStore } from '../../test/utils/inMemoryTauriStore';
import { initializeQueueStore, useQueueStore } from './queue.store';

vi.mock('@tauri-apps/plugin-store', async () => {
  const mod = await import('../../test/utils/inMemoryTauriStore');
  return { LazyStore: mod.LazyStore };
});

let uuidCounter = 0;
vi.mock('uuid', async () => {
  return {
    v4: () => `mock-uuid-${uuidCounter++}`,
  };
});

const createMockTrack = (title: string): Track => ({
  title,
  artists: [{ name: 'Test Artist', roles: ['primary'] }],
  source: { provider: 'test', id: title.toLowerCase() },
});

describe('useQueueStore', () => {
  beforeEach(() => {
    const date = new Date(2025, 1, 1);
    vi.setSystemTime(date);
    resetInMemoryTauriStore();
    useQueueStore.setState({
      items: [],
      currentIndex: 0,
      repeatMode: 'off',
      shuffleEnabled: false,
      isReady: false,
      isLoading: false,
    });
  });

  describe('initial state', () => {
    // Kinda redundant given the beforeEach, but we can ensure nothing weird happens on init
    it('starts empty and not ready', () => {
      const state = useQueueStore.getState();
      expect(state.items).toEqual([]);
      expect(state.currentIndex).toBe(0);
      expect(state.repeatMode).toBe('off');
      expect(state.shuffleEnabled).toBe(false);
      expect(state.isReady).toBe(false);
      expect(state.isLoading).toBe(false);
    });
  });

  describe('addToQueue', () => {
    it('appends tracks to the end', () => {
      const tracks = [createMockTrack('Track 1'), createMockTrack('Track 2')];
      useQueueStore.getState().addToQueue(tracks);

      const state = useQueueStore.getState();
      expect(state.items).toMatchSnapshot();
    });

    it('creates items with default status and metadata', () => {
      const track = createMockTrack('Test');
      useQueueStore.getState().addToQueue([track]);

      const state = useQueueStore.getState();
      expect(state.items).toMatchSnapshot();
    });
  });

  describe('addNext', () => {
    it('inserts tracks after current index', () => {
      const initial = [
        createMockTrack('Track 1'),
        createMockTrack('Track 2'),
        createMockTrack('Track 3'),
      ];
      useQueueStore.getState().addToQueue(initial);
      useQueueStore.setState({ currentIndex: 1 });

      useQueueStore.getState().addNext([createMockTrack('New Track')]);

      const state = useQueueStore.getState();
      expect(state.items).toHaveLength(4);
      expect(state.items[2].track.title).toBe('New Track');
      expect(state.items[3].track.title).toBe('Track 3');
    });
  });

  describe('addAt', () => {
    it('inserts tracks at specified index', () => {
      useQueueStore
        .getState()
        .addToQueue([createMockTrack('A'), createMockTrack('C')]);

      useQueueStore.getState().addAt([createMockTrack('B')], 1);

      const titles = useQueueStore
        .getState()
        .items.map((item) => item.track.title);
      expect(titles).toEqual(['A', 'B', 'C']);
    });

    it('adjusts currentIndex when inserting before it', () => {
      useQueueStore
        .getState()
        .addToQueue([
          createMockTrack('A'),
          createMockTrack('B'),
          createMockTrack('C'),
        ]);
      useQueueStore.setState({ currentIndex: 2 });

      useQueueStore.getState().addAt([createMockTrack('X')], 1);

      expect(useQueueStore.getState().currentIndex).toBe(3);
    });
  });

  describe('removeByIds', () => {
    it('removes items by their IDs', () => {
      useQueueStore
        .getState()
        .addToQueue([
          createMockTrack('A'),
          createMockTrack('B'),
          createMockTrack('C'),
        ]);
      const ids = useQueueStore.getState().items.map((item) => item.id);

      useQueueStore.getState().removeByIds([ids[1]]);

      const state = useQueueStore.getState();
      expect(state.items).toHaveLength(2);
      expect(state.items[0].track.title).toBe('A');
      expect(state.items[1].track.title).toBe('C');
    });

    it('adjusts currentIndex when removing items before it', () => {
      useQueueStore
        .getState()
        .addToQueue([
          createMockTrack('A'),
          createMockTrack('B'),
          createMockTrack('C'),
        ]);
      useQueueStore.setState({ currentIndex: 2 });
      const ids = useQueueStore.getState().items.map((item) => item.id);

      useQueueStore.getState().removeByIds([ids[0], ids[1]]);

      expect(useQueueStore.getState().currentIndex).toBe(0);
    });
  });

  describe('removeByIndices', () => {
    it('removes items by their indices', () => {
      useQueueStore
        .getState()
        .addToQueue([
          createMockTrack('A'),
          createMockTrack('B'),
          createMockTrack('C'),
        ]);

      useQueueStore.getState().removeByIndices([0, 2]);

      const state = useQueueStore.getState();
      expect(state.items).toHaveLength(1);
      expect(state.items[0].track.title).toBe('B');
    });
  });

  describe('clearQueue', () => {
    it('removes all items and resets index', () => {
      useQueueStore
        .getState()
        .addToQueue([createMockTrack('A'), createMockTrack('B')]);
      useQueueStore.setState({ currentIndex: 1 });

      useQueueStore.getState().clearQueue();

      const state = useQueueStore.getState();
      expect(state.items).toHaveLength(0);
      expect(state.currentIndex).toBe(0);
    });
  });

  describe('reorder', () => {
    it('moves item from one position to another and updates the current index', () => {
      useQueueStore
        .getState()
        .addToQueue([
          createMockTrack('A'),
          createMockTrack('B'),
          createMockTrack('C'),
        ]);

      useQueueStore.getState().reorder(0, 2);

      const titles = useQueueStore
        .getState()
        .items.map((item) => item.track.title);
      expect(titles).toEqual(['B', 'C', 'A']);
      expect(useQueueStore.getState().currentIndex).toBe(2);
    });

    it('adjusts currentIndex when moving item across it', () => {
      useQueueStore
        .getState()
        .addToQueue([
          createMockTrack('A'),
          createMockTrack('B'),
          createMockTrack('C'),
        ]);
      useQueueStore.setState({ currentIndex: 1 });

      useQueueStore.getState().reorder(0, 2);

      expect(useQueueStore.getState().currentIndex).toBe(0);
    });
  });

  describe('updateItemState', () => {
    it('updates specific item properties', () => {
      useQueueStore.getState().addToQueue([createMockTrack('Track')]);
      const itemId = useQueueStore.getState().items[0].id;

      useQueueStore.getState().updateItemState(itemId, {
        status: 'loading',
        activeStreamIndex: 1,
      });

      const item = useQueueStore.getState().items[0];
      expect(item.status).toBe('loading');
      expect(item.activeStreamIndex).toBe(1);
    });
  });

  describe('navigation', () => {
    beforeEach(() => {
      useQueueStore
        .getState()
        .addToQueue([
          createMockTrack('Track 1'),
          createMockTrack('Track 2'),
          createMockTrack('Track 3'),
        ]);
    });

    it('goToNext increments currentIndex', () => {
      useQueueStore.getState().goToNext();
      expect(useQueueStore.getState().currentIndex).toBe(1);
    });

    it('goToNext stops at end of queue', () => {
      useQueueStore.setState({ currentIndex: 2 });
      useQueueStore.getState().goToNext();
      expect(useQueueStore.getState().currentIndex).toBe(2);
    });

    it('goToPrevious decrements currentIndex', () => {
      useQueueStore.setState({ currentIndex: 2 });
      useQueueStore.getState().goToPrevious();
      expect(useQueueStore.getState().currentIndex).toBe(1);
    });

    it('goToPrevious stops at start of queue', () => {
      useQueueStore.getState().goToPrevious();
      expect(useQueueStore.getState().currentIndex).toBe(0);
    });

    it('goToIndex sets currentIndex to valid position', () => {
      useQueueStore.getState().goToIndex(2);
      expect(useQueueStore.getState().currentIndex).toBe(2);
    });

    it('goToIndex ignores out-of-bounds values', () => {
      useQueueStore.getState().goToIndex(10);
      expect(useQueueStore.getState().currentIndex).toBe(0);
    });

    it('goToId finds and navigates to item by ID', () => {
      const items = useQueueStore.getState().items;
      useQueueStore.getState().goToId(items[2].id);
      expect(useQueueStore.getState().currentIndex).toBe(2);
    });

    it('getCurrentItem returns the current queue item', () => {
      useQueueStore.setState({ currentIndex: 1 });
      const current = useQueueStore.getState().getCurrentItem();
      expect(current?.track.title).toBe('Track 2');
    });
  });

  describe('mode controls', () => {
    it('setRepeatMode updates repeat mode', () => {
      useQueueStore.getState().setRepeatMode('all');
      expect(useQueueStore.getState().repeatMode).toBe('all');
    });

    it('setShuffleEnabled toggles shuffle', () => {
      useQueueStore.getState().setShuffleEnabled(true);
      expect(useQueueStore.getState().shuffleEnabled).toBe(true);
    });
  });

  describe('persistence', () => {
    it('initializeQueueStore restores state from storage', async () => {
      const tracks = [createMockTrack('Track 1'), createMockTrack('Track 2')];
      useQueueStore.getState().addToQueue(tracks);
      useQueueStore.getState().goToIndex(1);
      useQueueStore.getState().setRepeatMode('all');

      await new Promise((resolve) => setTimeout(resolve, 10));

      useQueueStore.setState({
        items: [],
        currentIndex: 0,
        repeatMode: 'off',
        shuffleEnabled: false,
        isReady: false,
        isLoading: false,
      });

      await initializeQueueStore();

      const state = useQueueStore.getState();
      expect(state.items).toHaveLength(2);
      expect(state.currentIndex).toBe(1);
      expect(state.repeatMode).toBe('all');
      expect(state.isReady).toBe(true);
    });

    it('sanitizes out-of-bounds currentIndex on load', async () => {
      useQueueStore.getState().addToQueue([createMockTrack('Track 1')]);
      useQueueStore.setState({ currentIndex: 5 });

      useQueueStore.setState({
        items: [],
        currentIndex: 0,
        isReady: false,
        isLoading: false,
      });

      await initializeQueueStore();

      expect(useQueueStore.getState().currentIndex).toBe(0);
    });
  });
});
