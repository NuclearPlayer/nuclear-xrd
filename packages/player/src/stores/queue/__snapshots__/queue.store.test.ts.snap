// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useQueueStore > addToQueue > appends tracks to the end 1`] = `
[
  {
    "activeStreamIndex": 0,
    "addedAtIso": "2025-02-01T00:00:00.000Z",
    "failedStreamIndices": [],
    "id": "mock-uuid-0",
    "status": "idle",
    "track": {
      "artists": [
        {
          "name": "Test Artist",
          "roles": [
            "primary",
          ],
        },
      ],
      "source": {
        "id": "track 1",
        "provider": "test",
      },
      "title": "Track 1",
    },
  },
  {
    "activeStreamIndex": 0,
    "addedAtIso": "2025-02-01T00:00:00.000Z",
    "failedStreamIndices": [],
    "id": "mock-uuid-1",
    "status": "idle",
    "track": {
      "artists": [
        {
          "name": "Test Artist",
          "roles": [
            "primary",
          ],
        },
      ],
      "source": {
        "id": "track 2",
        "provider": "test",
      },
      "title": "Track 2",
    },
  },
]
`;

exports[`useQueueStore > addToQueue > creates items with default status and metadata 1`] = `
[
  {
    "activeStreamIndex": 0,
    "addedAtIso": "2025-02-01T00:00:00.000Z",
    "failedStreamIndices": [],
    "id": "mock-uuid-2",
    "status": "idle",
    "track": {
      "artists": [
        {
          "name": "Test Artist",
          "roles": [
            "primary",
          ],
        },
      ],
      "source": {
        "id": "test",
        "provider": "test",
      },
      "title": "Test",
    },
  },
]
`;
